File: elixir/dcarral/beer_song.exs
defmodule BeerSong do

  def lyrics do
    lyrics(99..0)
  end

  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(number) do
    case number do
      2 ->
        """
        2 bottles of beer on the wall, 2 bottles of beer.
        Take one down and pass it around, 1 bottle of beer on the wall.
        """
      1 ->
        """
        1 bottle of beer on the wall, 1 bottle of beer.
        Take it down and pass it around, no more bottles of beer on the wall.
        """
      0 ->
        """
        No more bottles of beer on the wall, no more bottles of beer.
        Go to the store and buy some more, 99 bottles of beer on the wall.
        """
      _ ->
        """
        #{number} bottles of beer on the wall, #{number} bottles of beer.
        Take one down and pass it around, #{number - 1} bottles of beer on the wall.
        """
    end
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range) do
    verses = Enum.map(range, &(verse(&1)))

    Enum.join(verses, "\n")
  end
end

File: java/dcarral/src/main/java/BeerSong.java
public class BeerSong {

    public String verse(int verseNumber) {
        if (verseNumber == 2) {
            return "2 bottles of beer on the wall, 2 bottles of beer.\n" +
                    "Take one down and pass it around, 1 bottle of beer on the wall.\n\n";
        } else if (verseNumber == 1) {
            return "1 bottle of beer on the wall, 1 bottle of beer.\n" +
                    "Take it down and pass it around, no more bottles of beer on the wall.\n\n";
        } else if (verseNumber == 0) {
            return "No more bottles of beer on the wall, no more bottles of beer.\n" +
                    "Go to the store and buy some more, 99 bottles of beer on the wall.\n\n";
        } else {
            return new StringBuilder().append(verseNumber)
                    .append(" bottles of beer on the wall, ")
                    .append(verseNumber)
                    .append(" bottles of beer.\n")
                    .append("Take one down and pass it around, ")
                    .append(verseNumber - 1)
                    .append(" bottles of beer on the wall.\n\n")
                    .toString();
        }

    }

    public String sing(int startingVerseNumber, int endingVerseNumber) {
        StringBuilder verses = new StringBuilder();

        for (int verseNumber = startingVerseNumber; verseNumber >= endingVerseNumber; verseNumber--) {
            String currentVerse = this.verse(verseNumber);
            verses.append(currentVerse);
        }

        return verses.toString();
    }

    public String singSong() {
        return this.sing(99, 0);
    }
}